### PR TITLE
improve Atomic Types documentation

### PR DESCRIPTION
**Pull request by orklah**
This PR synchronize the documentation of types in plugins_type_system.md and the documentation at the top of each class Type. I completed both when needed and created documentation for the rest. There's just 4 types I didn't know how to document

### CODE REVIEW

**Comment by orklah** [comment]
Those are the 4 types:
https://github.com/vimeo/psalm/pull/4735/files#diff-ef2714186d1cfc3aae07e2f422f9303e608f9b7973cccb28a1dd30875ffa090fR56

File: docs/running_psalm/plugins/plugins_type_system.md [hunk 1]
```diff
@@ -41,16 +41,40 @@ The classes are as follows:
...
+
+`TKeyOfClassConstant` - Represents an offset of a class constant array.
+
+`TTemplateIndexedAccess` - To be documented
+
+`TTemplateKeyOf` - To be documented
+
+`TTypeAlias` - To be documented
+
+`TValueOfClassConstant` - To be documented
```
Comment (weirdan) [inline docs/running_psalm/plugins/plugins_type_system.md:62]: Same as `TKeyOfClassConstant`, but for values, I suppose.

File: docs/running_psalm/plugins/plugins_type_system.md [hunk 2]
```diff
@@ -41,16 +41,40 @@ The classes are as follows:
...
+
+`TConditional` - Internal representation of a conditional return type in phpdoc. For example ($param1 is int ? int : string)
+
+`TIntMask` - Represents the type that is the result of a bitmask combination of its parameters. `int-mask<1, 2, 4>` corresponds to `1|2|3|4|5|6|7`
+
+`TIntMaskOf` - as above, but used with with a reference to constants in code`int-mask<MyClass::CLASS_CONSTANT_*>` will corresponds to `1|2|3|4|5|6|7` if there are three constant 1, 2 and 4
+
+`TKeyOfClassConstant` - Represents an offset of a class constant array.
+
+`TTemplateIndexedAccess` - To be documented
```
Comment (weirdan) [inline docs/running_psalm/plugins/plugins_type_system.md:56]: My guess would be that it corresponds to `T[K]` type mentioned in https://psalm.dev/docs/annotating_code/type_syntax/atomic_types/#magical-types

File: docs/running_psalm/plugins/plugins_type_system.md (hunk 1 above)
Comment (orklah) [reply docs/running_psalm/plugins/plugins_type_system.md:62]: Added, thanks

File: docs/running_psalm/plugins/plugins_type_system.md (hunk 2 above)
Comment (orklah) [reply docs/running_psalm/plugins/plugins_type_system.md:56]: mmh, not much description on that one. Not sure what I should add

Comment (weirdan) [reply docs/running_psalm/plugins/plugins_type_system.md:56]: See release notes for [Psalm 3.3.1](https://github.com/vimeo/psalm/releases/tag/3.3.1)

Comment (orklah) [reply docs/running_psalm/plugins/plugins_type_system.md:56]: Well, I added 
> Represents the type used when using TKeyOfClassConstant when the type of the class constant array is a template

Not sure I would be able to tell what it is if I didn't just checked the example but well, better than nothing

Comment (weirdan) [reply docs/running_psalm/plugins/plugins_type_system.md:56]: Typescript docs may provide some inspiration: https://www.typescriptlang.org/docs/handbook/advanced-types.html#index-types

**Comment by muglug** [comment]
Thanks, this is awesome!

**Pull request merged**